### PR TITLE
rgw/swift: preserve dashes/underscores in swift user metadata names

### PR DESCRIPTION
**Pull request by jzhu116-bloomberg**
Preserve dashes/underscores in swift user metadata names so it allows proper display and deletion.

Fixes: https://tracker.ceph.com/issues/51772
Signed-off-by: Jane Zhu <jzhu116@bloomberg.net>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cbodley** [comment]
thanks @jzhu116-bloomberg! does this change the format of the stored `RGW_ATTR_META_PREFIX` xattrs? have you been able to verify that user metadata written *before* this fix can still be accessed/removed after upgrading to the fixed version?

**Comment by jzhu116-bloomberg** [comment]
> thanks @jzhu116-bloomberg! does this change the format of the stored `RGW_ATTR_META_PREFIX` xattrs? have you been able to verify that user metadata written _before_ this fix can still be accessed/removed after upgrading to the fixed version?

No, it doesn't change the format of the stored `RGW_ATTR_META_PREFIX` xattrs.
Here are some tests that I've done on both "Quincy 17.2.0"  and the fixed version ("main + this PR").

After 
`$ swift -U test:tester -K testing -A http://xx.xx.xx.xx:8080/auth/1.0 post bucket -H X-Container-Meta-test_container_meta1:Meta1
$ swift -U test:tester -K testing -A http://xx.xx.xx.xx:8080/auth/1.0 post bucket -H X-Container-Meta-test-container-meta2:Meta2`

On both "Quincy 17.2.0" and "main + this PR", I have the following format. They are exactly the same on both versions.
`$ sudo rados -p dev-zone-bcc-master.rgw.meta -N root listxattr .bucket.meta.bucket:99e891b6-ee98-4625-bf42-acda8503bff0.5553.1
ceph.objclass.version
user.rgw.acl
user.rgw.x-amz-meta-test-container-meta2
user.rgw.x-amz-meta-test_container_meta1`

I also tested retrieving/removing the metadata set in version "Quincy 17.2.0" after I upgraded my cluster to the version "main + this PR"), and everything worked well.

However please note, with the fixed version, there is a display change in the output of "stat bucket". Please see the following examples. In the fixed version, the output reflects exactly the dashes/underscores in the metadata names, instead of always converting them to dashes in the older versions. 

On "Quincy 17.2.0",
`$ swift -U test:tester -K testing -A http://xx.xx.xx.xx:8080/auth/1.0 stat bucket
...
    Meta Test-Container-Meta2: Meta2
     Meta Test-Container-Meta: Meta1
...

On "main + this PR",
`$ swift -U test:tester -K testing -A http://xx.xx.xx.xx:8080/auth/1.0 stat bucket
...
    Meta Test-Container-Meta2: Meta2
     Meta Test_Container_Meta: Meta1
...
`

**Comment by mdw-at-linuxbox** [comment]
Couple of code nits: /1/ you can glue std::strings together using +, and /2/ you can ingest r-values (& eventually return them as your result) using "&&".  Using both of these should produce simpler better code.  (yes the original code didn't do that...)   { there's a stringstream use & an on-stack variable  C string allocation that I think you can eliminate with this... { and yes your stringstream is better than the original code snprintf...} }

Also, /3/ rather than having two slightly different copies of the lowercase-dash and camelcase-http functions I'd recommend adding an extra boolean flag to elicit the different behavior.  You can add a 'default' value for the parameter in .h if you only want the special behavior in a few places.

**Comment by alimaredia** [comment]
I ran this PR through teuthology a few weeks ago and the tests it said it fixed still failed: http://qa-proxy.ceph.com/teuthology/amaredia-2023-07-13_13:19:49-rgw:tempest-wip-rgw-swift-underscore-distro-default-smithi/7336183/teuthology.log

@mdw-at-linuxbox have you run tempest locally on the branch to see the tests that fail succeed?

**Comment by mattbenjamin** [comment]
agree, avoid stringstream if possible, maybe fmt can substitute?

**Comment by cbodley** [comment]
i just recently noticed that the swift API specifies that underscores *should* be converted to dashes:

> Note: Metadata keys (the name of the metadata) must be treated as case-insensitive at all times. These keys can contain ASCII 7-bit characters that are not control (0-31) characters, DEL, or a separator character, according to HTTP/1.1. The underscore character is silently converted to a hyphen.

this text shows up under both https://docs.openstack.org/api-ref/object-store/#create-container  and https://docs.openstack.org/api-ref/object-store/#create-update-or-delete-container-metadata

tempest test case [test_create_container_with_metadata_value](https://github.com/openstack/tempest/blob/master/tempest/api/object_storage/test_container_services.py#L70-L80) checks for this part and is failing with this PR

we used to handle this correctly for swift before https://github.com/ceph/ceph/pull/38737, which was a fix for the s3 side; however, those changes should have only applied to the s3 API

**Comment by jzhu116-bloomberg** [comment]
Checked in changes 
- based on code review
- to convert underscores to hyphens for swift metadata
- to allow deleting of the existing metadata

**Comment by prazumovsky** [comment]
Just to note: Quincy has the same issue so the potential backport required after https://tracker.ceph.com/issues/51772 will be fixed.

**Comment by prazumovsky** [comment]
Do we have any updates on this issue? We disabled corresponding tempest tests but want to see them working with a proper fix

**Comment by alimaredia** [comment]
@prazumovsky: When I run this PR through the tempests tests locally I still get failures for `tempest.api.object_storage.test_container_services.ContainerTest.test_create_container_with_remove_metadata_.*'`.  

I have a an additional fix (https://github.com/ceph/ceph-ci/commit/3999c5718ca7373e77753e7f3176790ba1641004) that got the tests to pass locally. I want to run it through the tempest tests in teuthology before it is added to this PR. I should have more information in the next day or two.

**Comment by prazumovsky** [comment]
thank you for the update!

**Comment by alimaredia** [comment]
The results of my teuthology runs verified that my commit was necessary:
Branch without my fix: https://github.com/ceph/ceph-ci/commits/wip-rgw-swift-preserve-dash-underscore-in-container-metadata
Run without my branch: http://pulpito.front.sepia.ceph.com/amaredia-2023-09-20_12:42:32-rgw:tempest-wip-rgw-swift-preserve-dash-underscore-in-container-metadata-distro-default-smithi/

Branch without my fix: https://github.com/ceph/ceph-ci/commits/wip-rgw-swift-preserve-dash-underscore-in-container-metadata-tempest-test-fix
Run without my branch: http://pulpito.front.sepia.ceph.com/amaredia-2023-09-20_12:42:01-rgw:tempest-wip-rgw-swift-preserve-dash-underscore-in-container-metadata-tempest-test-fix-distro-default-smithi/

**Comment by alimaredia** [comment]
@jzhu116-bloomberg @cbodley updated this PR with a commit that has the tempest tests passing. Will need someone else to review.

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -3433,9 +3433,6 @@ void RGWCreateBucket::execute(optional_yield y)
    * recover from a partial create by retrying it. */
   ldpp_dout(this, 20) << "rgw_create_bucket returned ret=" << op_ret << " bucket=" << s->bucket.get() << dendl;
 
-  if (op_ret)
-    return;
-
```
Comment (cbodley) [inline src/rgw/rgw_op.cc:3438]: discussed in the bug scrub - we want to return here on any fatal errors, but continue on if op_ret is either EEXIST or ERR_BUCKET_EXISTS

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -3433,7 +3433,7 @@ void RGWCreateBucket::execute(optional_yield y)
    * recover from a partial create by retrying it. */
   ldpp_dout(this, 20) << "rgw_create_bucket returned ret=" << op_ret << " bucket=" << s->bucket.get() << dendl;
 
-  if (op_ret)
+  if ((op_ret != 0 && op_ret != -EEXIST) || (op_ret != 0 && op_ret != -ERR_BUCKET_EXISTS))
```
Comment (cbodley) [inline src/rgw/rgw_op.cc:3436]:
```c++
if (op_ret < 0 && op_ret != EEXIST && op_ret != ERR_BUCKET_EXISTS)
```

**Comment by alimaredia** [comment]
http://pulpito.front.sepia.ceph.com/amaredia-2023-10-30_21:38:22-rgw-wip-rgw-swift-preserve-dash-underscore-in-container-metadata-distro-default-smithi/

**Comment by cbodley** [comment]
> http://pulpito.front.sepia.ceph.com/amaredia-2023-10-30_21:38:22-rgw-wip-rgw-swift-preserve-dash-underscore-in-container-metadata-distro-default-smithi/

@alimaredia please use public links: http://pulpito.ceph.com/amaredia-2023-10-30_21:38:22-rgw-wip-rgw-swift-preserve-dash-underscore-in-container-metadata-distro-default-smithi/

i see the relevent tempest tests are failing [[teuthology.log](http://qa-proxy.ceph.com/teuthology/amaredia-2023-10-30_21:38:22-rgw-wip-rgw-swift-preserve-dash-underscore-in-container-metadata-distro-default-smithi/7441439/teuthology.log)]:
tempest.api.object_storage.test_container_services.ContainerTest.test_create_container_with_remove_metadata_key
tempest.api.object_storage.test_container_services.ContainerTest.test_create_container_with_remove_metadata_value

**Comment by jzhu116-bloomberg** [comment]
> @alimaredia please use public links: http://pulpito.ceph.com/amaredia-2023-10-30_21:38:22-rgw-wip-rgw-swift-preserve-dash-underscore-in-container-metadata-distro-default-smithi/
> 
> i see the relevent tempest tests are failing [[teuthology.log](http://qa-proxy.ceph.com/teuthology/amaredia-2023-10-30_21:38:22-rgw-wip-rgw-swift-preserve-dash-underscore-in-container-metadata-distro-default-smithi/7441439/teuthology.log)]: tempest.api.object_storage.test_container_services.ContainerTest.test_create_container_with_remove_metadata_key tempest.api.object_storage.test_container_services.ContainerTest.test_create_container_with_remove_metadata_value

I did some more manual tests, and found out that POST request works fine, but not PUT request. The above 2 tempest tests are using PUT requests.

The very first PUT request to create a container with "X-Container-Meta-\*" header succeeds with status code 201, and the meta data is properly set. However the following PUT requests on the same container return with the status code 202 without handling the "X-Container-Meta-\*"/"X-Remove-Container-Meta-*" headers at all.

Please see below for some relevant log lines:
**The 1st PUT request:**
```
...
2023-11-01T03:25:59.513+0000 7f1798870700 20 HTTP_X_CONTAINER_META_TEST_CONTAINER_META=Meta1   
2023-11-01T03:25:59.513+0000 7f1798870700 20 REQUEST_METHOD=PUT                        
...
2023-11-01T03:25:59.513+0000 7f1798870700  1 ====== starting new request req=0x7f1696e6c790 =====
...
2023-11-01T03:25:59.518+0000 7f1794868700 10 req 17953498483730442036 0.005000074s swift:create_bucket x>> x-amz-meta-test-container-meta:Meta1
2023-11-01T03:25:59.518+0000 7f1794868700 10 req 17953498483730442036 0.005000074s swift:create_bucket user=test bucket=:bucket3[])
2023-11-01T03:25:59.518+0000 7f1794868700 10 req 17953498483730442036 0.005000074s swift:create_bucket cache get: name=default.rgw.meta+root+bucket3 : hit (negative entry)
...
2023-11-01T03:25:59.560+0000 7f178e05b700 20 req 17953498483730442036 0.047000695s swift:create_bucket rgw_create_bucket returned ret=0 bucket=:bucket3[43eefdcb-0adb-4c1d-9ce2-63ce3f37d3c9.4279.4])

2023-11-01T03:25:59.560+0000 7f178e05b700  2 req 17953498483730442036 0.047000695s swift:create_bucket completing
2023-11-01T03:25:59.560+0000 7f178e05b700 10 req 17953498483730442036 0.047000695s cache get: name=default.rgw.log++script.postrequest. : hit (negative entry)
2023-11-01T03:25:59.560+0000 7f178e05b700  2 req 17953498483730442036 0.047000695s swift:create_bucket op status=1900
2023-11-01T03:25:59.560+0000 7f178e05b700  2 req 17953498483730442036 0.047000695s swift:create_bucket http status=201
2023-11-01T03:25:59.560+0000 7f178e05b700  1 ====== req done req=0x7f1696e6c790 op status=1900 http_status=201 latency=0.047000695s ======  

````
**The 2nd PUT request on the same container:**
```
...                                                               
2023-11-01T03:25:59.566+0000 7f1792864700 20 HTTP_X_CONTAINER_META_TEST_CONTAINER_META=Meta2
2023-11-01T03:25:59.566+0000 7f1792864700 20 REQUEST_METHOD=PUT                           
...                                                          
2023-11-01T03:25:59.566+0000 7f1792864700  1 ====== starting new request req=0x7f1696e6c790 =====  
...
2023-11-01T03:25:59.568+0000 7f178684c700 10 req 14590071125625771467 0.002000030s swift:create_bucket x>> x-amz-meta-test-container-meta:Meta2
2023-11-01T03:25:59.568+0000 7f178684c700 10 req 14590071125625771467 0.002000030s swift:create_bucket user=test bucket=:bucket3[])
2023-11-01T03:25:59.568+0000 7f178684c700 10 req 14590071125625771467 0.002000030s swift:create_bucket cache get: name=default.rgw.meta+root+bucket3 : hit (requested=0x11, cached=0x17)
...
2023-11-01T03:25:59.575+0000 7f178b055700 20 req 14590071125625771467 0.009000134s swift:create_bucket rgw_create_bucket returned ret=-2013 bucket=:bucket3[43eefdcb-0adb-4c1d-9ce2-63ce3f37d3c9.4279.4])
2023-11-01T03:25:59.575+0000 7f178b055700  2 req 14590071125625771467 0.009000134s swift:create_bucket completing
2023-11-01T03:25:59.575+0000 7f178b055700 10 req 14590071125625771467 0.009000134s cache get: name=default.rgw.log++script.postrequest. : hit (negative entry)
2023-11-01T03:25:59.575+0000 7f178b055700  2 req 14590071125625771467 0.009000134s swift:create_bucket op status=1901
2023-11-01T03:25:59.575+0000 7f178b055700  2 req 14590071125625771467 0.009000134s swift:create_bucket http status=202
2023-11-01T03:25:59.576+0000 7f178b055700  1 ====== req done req=0x7f1696e6c790 op status=1901 http_status=202 latency=0.010000148s ======

````

What should be the correct behavior for a PUT request on an existing container in terms of custom metadata handling?
I see the following text in https://docs.openstack.org/api-ref/object-store/#create-container
> You do not need to check whether a container already exists before issuing a PUT operation because the operation is idempotent: It creates a container or updates an existing container, as appropriate.

Does this mean we should update all attributes specified in the PUT container request if the container exists?

**Comment by jzhu116-bloomberg** [comment]
> 
> Does this mean we should update all attributes specified in the PUT container request if the container exists?

Please ignore my question. I didn't read through thoroughly. Ali's fix actually answered it.
This doesn't work however. Somewhere else is wrong.

File: src/rgw/rgw_op.cc [hunk 3]
```diff
@@ -3433,7 +3433,7 @@ void RGWCreateBucket::execute(optional_yield y)
    * recover from a partial create by retrying it. */
   ldpp_dout(this, 20) << "rgw_create_bucket returned ret=" << op_ret << " bucket=" << s->bucket.get() << dendl;
 
-  if (op_ret)
+  if (op_ret < 0 && op_ret != EEXIST && op_ret != ERR_BUCKET_EXISTS)
```
Comment (jzhu116-bloomberg) [inline src/rgw/rgw_op.cc:3436]: shouldn't it be `if (op_ret < 0 && op_ret != -EEXIST && op_ret != -ERR_BUCKET_EXISTS)` ?

**Comment by jzhu116-bloomberg** [comment]
jenkins test make check

**Comment by jzhu116-bloomberg** [comment]
jenkins test api

**Comment by alimaredia** [comment]
https://pulpito.ceph.com/amaredia-2023-11-16_14:04:59-rgw-wip-rgw-swift-preserve-dash-underscore-in-container-metadata-distro-default-smithi/

**Comment by alimaredia** [comment]
jenkins test api

**Comment by prazumovsky** [comment]
Please actualizethis PR because the issue is still reproducable

**Comment by cbodley** [comment]
pushed a rebase that removed the merge commit. the tasks/tempest.yaml conflict was resolved automatically. will merge once checks pass

**Comment by prazumovsky** [comment]
Bugfix merged, great news! Hope to see it in reef and quincy.

**Comment by prazumovsky** [comment]
@jzhu116-bloomberg @cbodley 

Hello! Still hoping to see backport to Reef at least :)

**Pull request merged**